C7Engine/EntryPoints/MessageToEngine.cs

```diff
@@ -162,7 +172,7 @@ public override void process()
 			Player controller = EngineStorage.gameData.players.Find(p => p.guid == EngineStorage.uiControllerID);
 
 			foreach (MapUnit unit in controller.units) {
-				Console.WriteLine($"{unit}, path length: {unit.path?.PathLength() ?? 0}");
+				Console.WriteLine($"{unit}, path length: {unit.path.PathLength()}");
```

Comment:
I got a null reference crash on this line when testing a game that I had saved out then loaded back in. That null check on unit.path is still needed. Preferably we could prevent MapUnit.path from ever containing null in the first place. There is a TilePath.NONE object but evidently it's not being used somewhere it should be. In general, we're doing a halfhearted take on the empty object pattern which leaves us with the worst of both worlds as references to either null or an empty object appear unpredictably.

Also, after commenting out this line, I received another crash: "AI data not recognized" from PlayerAI.getAIForUnitStrategy.

Reply:
Yes, with regard to empty objects vs. null, it's pretty inconvenient in the code right now, since with the current lack of standard, in order to be sure you won't cause an exception, you basically need to check for null and the empty instance. And even when there is an empty instance, it's not consistent as to whether or not they can be treated as other instances, and in some cases they may contain null references. We should probably pick between null vs. empty objects and do some cleanup

Reply:
I think the AI data not recognised stems from the Json serialiser not being able to handle instances of an interface type. I haven't really touched the AI code, but is there a clear advantage to using interfaces over ABCs with only a single level of inheritance? tagging @QuintillusCFC if you have any thoughts on this. It seems like our options are either to write a bunch of code to serialise and deserialise instances of each interface type we want to save, swap out the interface for a class type, or reconstruct AI data at load / runtime.

Reply:
On the AI data: The main benefit of using the interface is that MapUnit.cs can store the data easily using the interface type.  With only concrete types, we'd either have to have unused storage spots for every type on MapUnit, or use whatever C#'s equivalent of Java's Object is... which would probably have the same serialization/deserialization challenges.

My inclination is to learn more about C# serialization, as we can't be the first project to want to serialize a concrete type and refer to it with an interface in our code.

As for null versus empty objects, I support the empty objects.  Null objects were one of the biggest pain points when working with my editor, and empty objects provide better readability as well.  It is true that we haven't done a good job of enforcing this (in part due to a hesitancy to enforce standards in the Wild West days), and that in some cases the empty objects weren't fully initialized.  I recall seeing the null TilePath at some point and wondering why the code wasn't just using TilePath.NONE .  This would be a good standardizing opportunity.